### ISSUE TITLE
pVACseq: Generating Variant Peptide FASTA and Key File. Error when processing proximal variant

### ISSUE THREAD

**Issue by tetsuro90** [griffithlab/pVACtools#688]
Hi, Thanks for the great software!

* pvactools version: docker images pVACtools:2.02
* Operating System: ubuntu 20.04

**Describe the bug**
When I ran pVACseq with the PHASED_PROXIMAL_VARIANTS_VCF option, I got the following error. This error occurred when processing the frameshift mutation in SIGLEC16, and I think it is due to the fact that SIGLEC16 is a polymorphic pseudogene. The version of VEP used for Somatic vcf and phsed vcf
are the same.

**Log Output**
```
Generating Variant Peptide FASTA and Key File
Error when processing proximal variant.
The wildtype amino acid for variant {my_somaticd_variant_index} with substring PASGLCHTQQPPSQAELDPV is different than expected.
Actual wildtype amino acid: Q
Wildtype amino acid of the proximal_variant: P
```

**My Consideration**
[SIGLEC16](http://asia.ensembl.org/Homo_sapiens/Gene/Summary?db=core;g=ENSG00000161643;r=19:49969600-49975819;t=ENST00000602139) has been registered as a polymorphic pseudogene in the reference sequence. The 21st amino acid is the stop codon（RAQRCCCCPCCCPCWGRGP*TRIPVTV........  [see here](http://asia.ensembl.org/Homo_sapiens/Transcript/Sequence_Protein?db=core;g=ENSG00000161643;r=19:49969600-49975819;t=ENST00000602139).). Therefore, the wildtype_amino_acid_sequence column in the {sample}.tsv file has 490 amino acids, but the frameshift_amino_acid_sequence column has only 20 amino acids (the position of my FS mutation was 371st).    
As a result, I believe that the problem occurs around lines 84 - 87 of the [get_frameshift_subsequence](https://github.com/griffithlab/pVACtools/blob/71cd12351cafc73057be87e8c717b1be59801ca1/lib/fasta_generator.py#L75) method.  The full_mutant_sequence is an array of length 20, but in my case, I think I sliced off 370 and beyond.  

```
        if self.downstream_sequence_length:
            mutant_subsequence_stop_position = position + self.downstream_sequence_length
            mutant_subsequence = full_mutant_sequence[start_position:mutant_subsequence_stop_position]
        else:
            mutant_subsequence = full_mutant_sequence[start_position:]
        left_flanking_sequence = full_mutant_sequence[start_position:position]
```

And I think that left_flanking_sequence becomes an array of zero length, mutation_start_position becomes zero in [line 200](https://github.com/griffithlab/pVACtools/blob/71cd12351cafc73057be87e8c717b1be59801ca1/lib/fasta_generator.py#L200), and finally there is a discrepancy in the add_proximal_variants method, which causes an error.  
Am I thinking correctly?
Also, is the only way to avoid the error is to remove the corresponding FS mutation from the vcf file?

Thanks,

**Comment by susannasiebert** [comment]
Thank you for reporting this issue. Would you be able to share two VCFs one with the somatic variant and the other with the matching proximal variant for this case with us so I can step through this case in the debugger?

**Comment by tetsuro90** [comment]
I am very sorry. It would be best if I could share it, but due to privacy reasons, I cannot share it.
The following information can be shared.

* GRCh38
* FS mutation chr19 49971925(SIGLEC16, at 371st, GC -> C)
* env proximal_variants SIGLEC16 368, 379, 380, missense

**Comment by susannasiebert** [comment]
I'm really not sure that I will be able to debug this issue without some example VCFs. Would you be able to just provide anonymized VCFs with the minimal information you already gave use? Basically remove all other variants, remove all identifying FILTER, INFO, and FORMAT fields and their headers etc. The only thing we really need are the CSQ, the WildtypeProtein, and the FrameshiftSequence fields.

**Comment by tetsuro90** [comment]
Thanks for the suggestion. I will look into whether that is possible. I'll need a little more time to think about this.

**Comment by tetsuro90** [comment]
@susannasiebert 
I created vcf files from which I extracted only the necessary information. I have confirmed that I can reproduce the error using these files.

I would like to e-mail you directly, which address should I send it to?

**Comment by susannasiebert** [comment]
You can email them to susanna.kiwala@wustl.edu

**Comment by tetsuro90** [comment]
Thank you.
I just sent you an email.

**Comment by susannasiebert** [comment]
I received your files and was able to reproduce the error. I'm working on a fix. May I include the example VCF you sent me as a test file for our unit tests? I would only include the somatic example VCF (my_sample_extracted.vcf), not the phased proximal one. I would also remove all of the allele count, gene expression, etc.

**Comment by malachig** [comment]
Hi @tetsuro90 .  Thanks for reporting this issue.  We have a working theory that this has to do with this variant occurring in a special, very rare class of pseudogene: **Polymorphic pseudogene**.  These are not handled like other pseudogenes in Ensembl.  We have a plan to address them.  

This article has a nice description of this gene biotype.
https://academic.oup.com/nar/article/46/14/7070/5047265

Currently, these are _only_ pseudogene biotype in Ensembl GTFs that define CDS features.  They represent a tiny fraction of all pseudogene transcripts (and an even tinier fraction of all gene transcripts).

The bottom line is that these are probably not a reliable source for neoantigens and it would be safest to skip them because we can generally not know if the sequence is truly a pseudogene, or in some individuals may produce functional proteins.  According to the reference sequence they are pseudogenes, and that is the information we have to extract protein sequences.

If we had high fidelity full length cDNA sequence data readily available we might think about this problem differently.  But for now we think this is the best solution.

**Comment by tetsuro90** [comment]
@susannasiebert 
You can use it as a test file. Thank you.

@malachig 
Hi, malachig. Thank you for the detailed explanation. I will also read the paper you linked to.
 It was clear to me that without high fidelity full length cDNA sequence data, no reasonable antigen prediction can be made.


Let me ask you another question. I was reading the source code of class ProximalVariant, but there is something I don't understand.  Does it support antigen prediction if the somatic FS and proximal FS are in a neighborhood (or if two or more somatic FSs are in a neighborhood)?

**Comment by susannasiebert** [comment]
Thank you. I will try to get an updated version of pVACtools out this week.

Unfortunately, only missense proximal variants are currently supported so if the proximal variant is a frameshift or an in-frame indel, we don't currently support incorporating those. You may find our proximal variant paper of interest on this topic: https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6309579/. We hope to add support for more proximal variant types in the future.

**Comment by tetsuro90** [comment]
I think incorporating FS proximal variant is very difficult work.
I will read this paper. Thank you.
